lib/workers/repository/update/branch/index.ts

```diff
@@ -463,7 +463,8 @@ export async function processBranch(
       );
       config.reuseExistingBranch = false;
     } else {
-      config = { ...config, ...(await shouldReuseExistingBranch(config)) };
+      const parentConfig = await shouldReuseExistingBranch(config);
+      config = { ...config, ...parentConfig };
```

Comment:
How is this different?

Reply:
Before, `{ ...config, /* ... */ }` was first included in the new object before `await shouldReuseExistingBranch(config)` had resolved and internally updated `config`, so the updated `config.rebaseWhen` value was not included in the new object and thus not passed through. Now, `await shouldReuseExistingBranch(config)` gets resolved first and `config.rebaseWhen` gets updated, then the two configs `{ ...config, ...parentConfig }` get merged where `config` has already been updated.

Reply:
See also [what `config = { ...config, ...parentConfig };` is transpiled to](https://www.typescriptlang.org/play/?#code/MYewdgZglg5gBAXjgbzgOg6SsA06MAOAhgE4CmYALgMLjTwC+A3EA), this makes it clear:

```js
config = Object.assign(Object.assign({}, config), parentConfig);
```

Reply:
OK, so this is because we added a mutation in shouldReuseExistingBranch()? I hadn't realized this was what I was proposing. Can we instead _return_ the changed rebaseWhen?

Reply:
Not without extending the [`ParentBranch` type](https://github.com/renovatebot/renovate/blob/26337ac124f4dd1885a44ed65822f2944ad69f3d/lib/workers/repository/update/branch/reuse.ts#L7-L11). `shouldReuseExistingBranch(...)` resolves to a `ParentBranch` object which is not a `BranchConfig` and does not have the `rebaseWhen` key.

Reply:
I think this is preferable and more easier to maintain in future than the current mutation

Reply:
Shall I simply add `rebaseWhen` to the `ParentBranch` type although it doesn't seem semantically related? Or shall I merge the config with the current return value of `shouldReuseExistingBranch(...)` and change its return type to `BranchConfig`?

Reply:
yes, merge inside that function and return the new config. change return type too of cause